telemetry/pendo/class-pendo-javascript-library.php

```diff
@@ -0,0 +1,167 @@
+ */
+class Pendo_JavaScript_Library {
+	/**
+	 * The Pendo API key.
+	 *
+	 * @var string
+	 */
+	private string|null $api_key;
+
+	/**
+	 * The global variable name for the localized Pendo data.
+	 *
+	 * @var string
+	 */
+	private string $browser_global = 'VIPPendo';
+
+	/**
+	 * Singleton instance.
+	 *
+	 * @var Pendo_JavaScript_Library
+	 */
+	private static $instance;
+
+	/**
+	 * Constructor.
+	 */
+	private function __construct( string $api_key ) {
+		$this->api_key = $api_key;
+	}
+
+	/**
+	 * Standard singleton class except the caller does not need access to the
+	 * instance, so we name it `init`. Do not call this method directly; instead
+	 * use `Pendo::enable_javascript_library()`.
+	 *
+	 * @param string $api_key The Pendo snippet API key.
+	 * @return bool|WP_Error True on success or WP_Error if any error occured.
+	 */
+	public static function init( string|null $api_key = null ): bool|WP_Error {
+		// Already initialized? Return early.
+		if ( self::$instance instanceof Pendo_JavaScript_Library ) {
+			return true;
+		}
+
+		if ( empty( $api_key ) ) {
+			$message = __( 'Pendo snippet API key is not defined', 'vip-telemetry' );
+			log2logstash( [
+				'severity' => 'error',
+				'feature'  => 'telemetry',
+				'message'  => $message,
+			] );
+			return new WP_Error( 'pendo_snippet_api_key_not_defined', $message );
+		}
+
+		// We currently do not allow customization of the initialization data to
+		// minimize the risk of collisions with standard property names.
+
+		self::$instance = new Pendo_JavaScript_Library( $api_key );
+
+		// Use a high priority value to enqueue this late.
+		add_action( 'admin_enqueue_scripts', [ self::$instance, 'enqueue_scripts' ], 99, 0 );
+
+		return true;
+	}
+
+	/**
+	 * Enqueue the Pendo client library script on supported pages.
+	 */
+	public function enqueue_scripts(): void {
+		if ( true !== self::should_enqueue_script() ) {
+			return;
+		}
+
+		$initialization_data = $this->get_initialization_data();
+		if ( is_wp_error( $initialization_data ) ) {
+			return;
+		}
+
+		$script_handle = 'vip-pendo-agent-script';
+		$script_url    = plugins_url( '/js/pendo-agent.js', __FILE__ );
+
+		$localized_data = [
+			'apiKey'   => $this->api_key,
+			'initData' => $initialization_data,
+		];
+
+		wp_enqueue_script( $script_handle, $script_url, [], '0.4', true );
+		wp_localize_script( $script_handle, $this->browser_global, $localized_data );
+	}
+
+	private function get_initialization_data(): array|WP_Error {
+		$event_properties = get_base_properties_of_pendo_track_event();
+		$user_properties  = get_base_properties_of_pendo_user();
+
+		// A null response indicates that the user is not logged-in or we otherwise
+		// have no meaningful way to attribute the event.
+		if ( empty( $user_properties ) ) {
+			$message = __( 'User is not logged-in', 'vip-telemetry' );
+			log2logstash( [
+				'severity' => 'error',
+				'feature'  => 'telemetry',
+				'message'  => $message,
+			] );
```

Comment:
why do we want to log this?

Reply:
All of the `log2logstash` calls highlight problems that should not be possible if the code is being called correctly. I mostly following the pattern of existing "doing it wrong" logging in the Tracks classes ([example](https://github.com/Automattic/vip-go-mu-plugins/blob/f666cb9c0e899a8c970562a1fad75b6e5433b83a/telemetry/tracks/class-tracks-event.php#L192)).

This specific call indicates that the defensive check [here](https://github.com/Automattic/vip-go-mu-plugins/blob/f666cb9c0e899a8c970562a1fad75b6e5433b83a/telemetry/pendo/class-pendo.php#L133) somehow failed or (more likely) someone is directly using the `Pendo_JavaScript_Library` class against advice and forgetting to pass the API key.

Reply:
To actually answer your question: Pendo Telemetry is restricted to users logged in to the WP Admin dashboard. This error likely indicates that the library is being called in a context that it shouldn't be, or attribution is unexpectedly not working.